common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java

```diff
@@ -0,0 +1,315 @@
+import com.thoughtworks.go.domain.MaterialInstance;
+import com.thoughtworks.go.domain.MaterialRevision;
+import com.thoughtworks.go.domain.config.ConfigurationProperty;
+import com.thoughtworks.go.domain.materials.*;
+import com.thoughtworks.go.domain.materials.scm.PluggableSCMMaterialInstance;
+import com.thoughtworks.go.domain.materials.scm.PluggableSCMMaterialRevision;
+import com.thoughtworks.go.domain.scm.SCM;
+import com.thoughtworks.go.util.StringUtil;
+import com.thoughtworks.go.util.command.EnvironmentVariableContext;
+import com.thoughtworks.go.util.command.ProcessOutputStreamConsumer;
+import com.thoughtworks.go.util.json.JsonHelper;
+import com.thoughtworks.go.util.json.JsonMap;
+
+import java.io.File;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.thoughtworks.go.util.command.EnvironmentVariableContext.escapeEnvironmentVariable;
+import static java.lang.String.format;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.apache.commons.lang.StringUtils.upperCase;
+
+public class PluggableSCMMaterial extends AbstractMaterial {
+    public static final String TYPE = "PluggableSCMMaterial";
+
+    private String scmId;
+
+    @Expose
+    @SerializedName("scm")
+    private SCM scmConfig;
+
+    private String folder;
+
+    private Filter filter;
+
+    private String fingerprint;
+
+    public PluggableSCMMaterial() {
+        super(TYPE);
+    }
+
+    public PluggableSCMMaterial(String scmId) {
+        this();
+        this.scmId = scmId;
+    }
+
+    public PluggableSCMMaterial(PluggableSCMMaterialConfig config) {
+        this();
+        this.name = config.getName();
+        this.scmId = config.getSCMId();
+        this.scmConfig = config.getSCMConfig();
+        this.folder = config.getFolder();
+        this.filter = config.filter();
+    }
+
+    @Override
+    public MaterialConfig config() {
+        return new PluggableSCMMaterialConfig(name, scmConfig, folder, filter);
+    }
+
+    @Override
+    public Class getInstanceType() {
+        return PluggableSCMMaterialInstance.class;
+    }
+
+    @Override
+    public MaterialInstance createMaterialInstance() {
+        return new PluggableSCMMaterialInstance(JsonHelper.toJsonString(this), UUID.randomUUID().toString());
+    }
+
+    public SCM getScmConfig() {
+        return scmConfig;
+    }
+
+    public void setSCMConfig(SCM scmConfig) {
+        this.scmConfig = scmConfig;
+    }
+
+    @Override
+    public String getFolder() {
+        return folder;
+    }
+
+    public void setFolder(String folder) {
+        this.folder = folder;
+    }
+
+    public Filter getFilter() {
+        return filter;
+    }
+
+    public void setFilter(Filter filter) {
+        this.filter = filter;
+    }
+
+    public String getPluginId() {
+        return getScmConfig().getPluginConfiguration().getId();
```

Comment:
on same lines as [packagematerial](https://github.com/gocd/gocd/blob/7958ac8b613031660ee00f0591af792c06c0f74c/common/src/com/thoughtworks/go/config/materials/PackageMaterial.java#L91). I think packagematerial isn't valid until packgedefinition is set & hence the same assumption can be made? @pshivana what do you think?